### ISSUE TITLE
check_nt CPULOAD returns extremly high CPU Load above 100% 

### ISSUE THREAD

**Issue by ferox147** [mickem/nscp#824]
### Issue and Steps to Reproduce

Running "check_nt" CPULOAD on a physically installed Windows Server 2022 Standard returns e.g.: CPU Load 18446744017345409802% (5 min average)

```
'./icinga2/addons/check_nt' '-H' '$hostadress$' '-l' '5,80,90' '-p' '1248' '-s' '$secret$' '-v' 'CPULOAD'
```

### Expected Behavior

In fact, a correct value between 0 and 100 % should be returned.

### Actual Behavior

Unfortunately, an extremely high, unrealistic value is returned, which does not reflect the behavior of the system.
(18446744017345409802% | 5 min avg)

### Details

* NSClient++ version: NSClient++ 0.5.2.41 | 2018-04-26
* OS and Version: Windows Server 2022 Standard
* Checking from: Icinga
* Checking with: check_nt

### Additional Details

CPU: 2x 8-core / 16-threads: 32 logical processors

**Comment by ferox147** [comment]
Also tested with check_nrpe command, I get the following negative percentage result:

OK: CPU load is ok.|'total 5m'=-3041937069%;80;90 'total 1m'=0%;80;90 'total 5s'=0%;80;90

**Comment by Marco-Jacobs** [comment]
It's the same here.
We monitor around 100 hosts with NRPE.
This behaviour only occurs sometimes and only on 5 hosts.

- The NSC versions are different 0.5.x versions
- The operating systems of the target computers are different Windows Server versions

**Comment by dneuhaeuser** [comment]
same here with one 2022 server, which has 56 cores.

Maybe related to core count?
I don't see this problem on 2022 servers which have 32 or less cores.

**Comment by wopfel** [comment]
Maybe related to #780.

**Comment by wopfel** [comment]
To get at least any valid data, I'm using the following check. It's fine for getting just the CPU utilization, until this issue is fixed.

```
    vars.nrpe_arguments = [
        "counter=\\Processor(_Total)\\% Processor Time",
        "perf-syntax=processor_time"
    ]
    vars.nrpe_command = "check_pdh"
```

**Comment by mickem** [comment]
I have a new version out soon (ish) which will possibly solve this., Unfortunately my machines only has 12 cores so I cannot verify.

It does add a debug log message which will display number of detected cores so at least we can isolate what the issue is:
```
L  w32system Number of cores detected (new) as 12
```

I also adds a shorthand `cores` to `check_cpu` to simplify displaying load per core:
```
check_cpu cores
L        cli OK: OK: CPU load is ok.
L        cli  Performance data: 'total 5m'=0%;80;90 'core 0 1m'=4%;80;90 'core 1 1m'=2%;80;90 'core 10 1m'=4%;80;90 'core 11 1m'=2%;80;90 'core 2 1m'=4%;80;90 'core 3 1m'=2%;80;90 'core 4 1m'=3%;80;90 'core 5 1m'=2%;80;90 'core 6 1m'=3%;80;90 'core 7 1m'=2%;80;90 'core 8 1m'=2%;80;90 'core 9 1m'=2%;80;90 'total 1m'=3%;80;90 'core 0 5s'=12%;80;90 'core 1 5s'=6%;80;90 'core 10 5s'=13%;80;90 'core 11 5s'=6%;80;90 'core 2 5s'=11%;80;90 'core 3 5s'=9%;80;90 'core 4 5s'=7%;80;90 'core 5 5s'=3%;80;90 'core 6 5s'=10%;80;90 'core 7 5s'=5%;80;90 'core 8 5s'=5%;80;90 'core 9 5s'=4%;80;90 'total 5s'=7%;80;90
```
 You should be able to download it here: https://github.com/mickem/nscp/actions/runs/13099468273?pr=880 in a bit...
Please test this and let me know what you get on a machine with more than 16 cores

**Comment by dneuhaeuser** [comment]
I have installed this version on a server 2022 now. 
Unfortunately the problem is still occuring:
`CRITICAL:  CPU Load 2528280749% (15 min average) `

btw: this value is not reported constantly. 
about 80% of the time I get "OK:  CPU Load 0% (15 min average)" while taskmanager is showing loads around 5-10%

just let me know when I can help debug this further...

**Comment by mickem** [comment]
Yes, I would like to see is the `L  w32system Number of cores detected (new) as 12` log message matches the number of cores in the machines.
Also I would like to see if `check_cpu cores` will provide reasonable values for the various cores.
This will hopefully help me pin down the cause which I hope is either wrong number of cores and/or wrong summation of core load.

**Comment by dneuhaeuser** [comment]
OK, log reports 56 cores which corresponds to the 56 physical cores of the machine:
`2025-02-10 17:45:21: info:D:\a\nscp\nscp\include\win_sysinfo\win_sysinfo.cpp:372: Number of cores detected (new) as 56`

this is the output of **check_nrpe -H xxx -c check_cpu -a cores**:
`OK: CPU load is ok.|'core 0 5m'=-1674596965%;80;90 'core 1 5m'=-5390829180%;80;90 'core 10 5m'=-5317980179%;80;90 'core 11 5m'=-5317980244%;80;90 'core 12 5m'=-5317980181%;80;90 'core 13 5m'=-5317980242%;80;90 'core 14 5m'=-5390828961%;80;90 'core 15 5m'=-5390829219%;80;90 'core 16 5m'=-5390829025%;80;90 'core 17 5m'=-5317980240%;80;90 'core 18 5m'=-5317980233%;80;90 'core 19 5m'=-5317980246%;80;90 'core 2 5m'=-5390829085%;80;90 'core 20 5m'=-5247073925%;80;90 'core 21 5m'=-5247073898%;80;90 'core 22 5m'=-5317980225%;80;90 'core 23 5m'=-5317980235%;80;90 'core 24 5m'=-5317980240%;80;90 'core 25 5m'=-5317980240%;80;90 'core 26 5m'=-5317980230%;80;90 'core 27 5m'=-5247073902%;80;90 'core 28 5m'=-5317980229%;80;90 'core 29 5m'=-5317980237%;80;90 'core 3 5m'=-5390829225%;80;90 'core 30 5m'=-5317980225%;80;90 'core 31 5m'=-5317980243%;80;90 'core 32 5m'=-5247073911%;80;90 'core 33 5m'=-5247073883%;80;90 'core 34 5m'=-5317980236%;80;90 'core 35 5m'=-5317980242%;80;90 'core 36 5m'=-5317980237%;80;90 'core 37 5m'=-5247073902%;80;90`
output seems to be cut off since not all cores are visible?

**Comment by mickem** [comment]
Yes, NRPE has a payload limit on 1K in version 2. But this is good enough and we can clearly see that we have negative values per core... So the problem is elsewhere... I will dig further...

**Comment by mickem** [comment]
Exploring another avenue here: https://github.com/mickem/nscp/actions/runs/13344141188

This dies two things: 
1. Change the match a bit (before I stored derived numbers for delta calculation) and now I derive the values from deltas of the original numbers, In theory this should not matter, but if the number are "incorrect" the error could compound so lets see if this is better
2. Add an option to switch to another API. This API does not get the CPU load per core so only retrieves the totals.

So it would be nice if someone with these issue could give this a spina nd let me know:

### 1, If this new calculation makes a difference?

Output of:
```
check_nrpe -H xxx -c check_cpu -a cores
```


### 2, If the other API produces better results.

Output of:
```
check_nrpe -H xxx -c check_cpu -a cores
```
After changing settings in nsclient.ini like so:
```
[/settings/system/windows]
fetch core loads = false
```

### 3, What kind of machine is required to reproduce this?

Is there some specific CPU and/or could it be reproduced in a cloud setting like on gcp/aws or azure if so what machine type?

**Comment by dneuhaeuser** [comment]
Here are my results with the new build:

1)
# ./check_nrpe -H x.x.x.x -c check_cpu -a cores
OK: CPU load is ok.|'core 0 5m'=-1230399784071%;80;90 'core 1 5m'=-3353979059775%;80;90 'core 10 5m'=-3318620843444%;80;90 'core 11 5m'=-3318620825718%;80;90 'core 12 5m'=-3318620844169%;80;90 'core 13 5m'=-3318620826574%;80;90 'core 14 5m'=-3353979084860%;80;90 'core 15 5m'=-3353979055224%;80;90 'core 16 5m'=-3353979077562%;80;90 'core 17 5m'=-3318620827683%;80;90 'core 18 5m'=-3318620828299%;80;90 'core 19 5m'=-3318620825214%;80;90 'core 2 5m'=-3353979070123%;80;90 'core 20 5m'=-3284000406314%;80;90 'core 21 5m'=-3284000392829%;80;90 'core 22 5m'=-3318620836237%;80;90 'core 23 5m'=-3318620828690%;80;90 'core 24 5m'=-3318620826423%;80;90 'core 25 5m'=-3318620826478%;80;90 'core 26 5m'=-3318620830231%;80;90 'core 27 5m'=-3284000393695%;80;90 'core 28 5m'=-3318620831193%;80;90 'core 29 5m'=-3318620829076%;80;90 'core 3 5m'=-3353979055044%;80;90 'core 30 5m'=-3318620833088%;80;90 'core 31 5m'=-3318620826105%;80;90 'core 32 5m'=-3284000397457%;80;90 'core 33 5m'=-3284000387646%;80;90 'core 34 5m'=-3318620829297%;80;90

2)
# ./check_nrpe -H x.x.x.x -c check_cpu -a cores
CRITICAL: 5m: 9519364643%, 1m: 47596823217%|'total 5m'=9519364643%;80;90 'total 1m'=47596823217%;80;90 'total 5s'=0%;80;90

3)
In my case it is a Supermicro Server with XEON Gold 5520.

**Comment by mickem** [comment]
Hmm... strange that 5s was 0 and 1m not maybe there is something amiss with the rrd buffer?...

Could you also try:
 ```
./check_nrpe -H x.x.x.x -c check_cpu -a time=1s
```
And run that both after service restart and after like 1-2 minutes?


As for the CPU, right, that unfortunately outside my budget :)
I tried looking for normal CPU:s and failed to find any with that many cores, and given the cost of that one I now understand why...

**Comment by dneuhaeuser** [comment]
`./check_nrpe -H x.x.x.x -c check_cpu -a time=1s`

after service restart:
`OK: CPU load is ok.|'total 1s'=2%;80;90`

after 1m
`CRITICAL: 1s: 746117%|'total 1s'=746117%;80;90`

after 2m
`OK: CPU load is ok.|'total 1s'=0%;80;90`

after 4m
`CRITICAL: 1s: 4068198534323%|'total 1s'=4068198534323%;80;90`

so it's flapping. but it's doing this every few minutes - not only after service restart.

**Comment by mickem** [comment]
Yet another attempt at isolating the issue. The build should be here in an hour or so: https://github.com/mickem/nscp/actions/runs/13348823268 (unless the build fails in which case it will be there tomorrow)
This time I have added logging of the values parsed (and tweaked the code a bit).

> Please don't leave it running as it will spam the log indefinitely every 1 second.

In the log you should see:

![Image](https://github.com/user-attachments/assets/3e5cd757-c9c3-4360-be65-3df2eedfe1b6)

Which is the values read and pushed, if these values are not "sensible" we know the issue is in the system API, if these values are sensible, the issue is in the collection.

**Comment by dneuhaeuser** [comment]
OK, I had this running for about an hour and did **not** see these extreme values anymore.


output in idle:
`OK: CPU load is ok.|'total 5m'=0%;80;90 'total 1m'=0%;80;90 'total 5s'=0%;80;90`

log snippet:
```
2025-02-16 01:26:30: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 99,478022, kernel: 0,164835, load: 0,521978
2025-02-16 01:26:31: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 99,532967, kernel: 0,219780, load: 0,467033
2025-02-16 01:26:32: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 99,776786, kernel: 0,083705, load: 0,223214
2025-02-16 01:26:33: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 99,780220, kernel: 0,109890, load: 0,219780
2025-02-16 01:26:34: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 99,804688, kernel: 0,139509, load: 0,195313
```
(core count = 0 ??)


I also put some load on the machine to check the values:
taskmanager showed a constant load of 30% (+/- 1%)

output:
`OK: CPU load is ok.|'total 5m'=28%;80;90 'total 1m'=37%;80;90 'total 5s'=37%;80;90`

log:
```
2025-02-16 01:29:39: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 62,165179, kernel: 34,123884, load: 37,834821
2025-02-16 01:29:40: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 62,362637, kernel: 34,010989, load: 37,637363
2025-02-16 01:29:41: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 62,060440, kernel: 34,148352, load: 37,939560
2025-02-16 01:29:42: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 62,193080, kernel: 34,123884, load: 37,806920
2025-02-16 01:29:43: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 62,417582, kernel: 33,956044, load: 37,582418
```

another test load:
taskmanager 55% (constantly) -> check_cpu: 50%

is this inaccuracy to be expected or is it possible that not all cores are taken into account yet?

**Comment by mickem** [comment]
So the "original problem" has been resolved then i.e. the results no longer are large negative values?
Thats good news... The real problem was a missing copy constructor.

The `0 cores` is likely due to you still having the following config:
```
[/settings/system/windows]
fetch core loads = false
```

As with that set to false we don't look for load per core and instead hardcode core to 0 and look for the over all load.
So we likely want to see if it also works correctly if you set:
```
[/settings/system/windows]
fetch core loads = true
```
(or comment the line out as `true` is the default).

As for the cpu discrepancy from what I can tell the two APIs produce different results as they measure things different so there could be some discrepancy between them.
That said the documentation for the APIs are  abit strange so I am not 100% sure I have done the math correctly when I tried "stuff".
So if we have solved the original problem, I shall review that to ensure the values are correct. I also had a few % discrepancy on my machine.

**Comment by dneuhaeuser** [comment]
I set **fetch core loads = true** now and the core count in log shows as expected:

`2025-02-16 19:17:37: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 56, idle: 99,135045, kernel: 0,334821, load: 0,864955
`

But the large load values reappeared.
I also have a feeling that my first 1 hour test was too short.
When looking at the history, there were always timespans (a few hours) where the large values did not show up.

So I let this run for a few more hours now and just catched this:
```
2025-02-16 20:19:35: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 56, idle: 98,051948, kernel: 0,838745, load: 1,948052
2025-02-16 20:19:36: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 56, idle: 957360,323339, kernel: 140108775510383,984375, load: -957260,323339
2025-02-16 20:19:37: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 56, idle: 159082015926245,468750, kernel: 11395672340942,953125, load: -159082015926145,468750
2025-02-16 20:19:38: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 56, idle: 98,353795, kernel: 0,474330, load: 1,646205
```

**./check_nrpe -H x.x.x.x -c check_cpu -a cores**
```
OK: CPU load is ok.|'core 0 5m'=-168175443370%;80;90 'core 1 5m'=-546570194496%;80;90 'core 10 5m'=-539082931509%;80;90 'core 11 5m'=-539082931718%;80;90 'core 12 5m'=-539082931496%;80;90 'core 13 5m'=-539082931711%;80;90 'core 14 5m'=-546570194002%;80;90 'core 15 5m'=-546570194585%;80;90 'core 16 5m'=-546570194152%;80;90 'core 17 5m'=-539082931696%;80;90 'core 18 5m'=-539082931689%;80;90 'core 19 5m'=-539082931725%;80;90 'core 2 5m'=-546570194311%;80;90 'core 20 5m'=-531798027212%;80;90 'core 21 5m'=-531798027286%;80;90 'core 22 5m'=-539082931591%;80;90 'core 23 5m'=-539082931683%;80;90 'core 24 5m'=-539082931710%;80;90 'core 25 5m'=-539082931710%;80;90 'core 26 5m'=-539082931664%;80;90 'core 27 5m'=-531798027282%;80;90 'core 28 5m'=-539082931654%;80;90 'core 29 5m'=-539082931678%;80;90 'core 3 5m'=-546570194584%;80;90 'core 30 5m'=-539082931630%;80;90 'core 31 5m'=-539082931715%;80;90 'core 32 5m'=-531798027259%;80;90 'core 33 5m'=-531798027316%;80;90 'core 34 5m'=-539082931673%;80;90 'core 35 5m'=-539082931702%;80;90
```

**Comment by mickem** [comment]
But as long as you set cores to false it works reliably?

**Comment by mickem** [comment]
I have released a pre-release with this change (without the log) if anyone wants to see if this new API works better.
Install this version: https://github.com/mickem/nscp/releases/tag/0.6.11

And configure the following:
```
[/settings/system/windows]
fetch core loads = false
```

**Comment by dneuhaeuser** [comment]
> But as long as you set `fetch core loads = false` to false it works reliably?

I tested this again today and unfortunately the problem occured again after ~2 hours:

```
2025-02-17 15:10:39: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 97,835498, kernel: 0,487013, load: 2,164502
2025-02-17 15:10:40: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 1069976,197680, kernel: 2976781348191,178223, load: 2976780626335,126953
2025-02-17 15:10:41: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 2767443029643,392578, kernel: 323547,398031, load: 994641,819034
2025-02-17 15:10:42: info:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:294: CPU Load: cores: 0, idle: 98,353795, kernel: 0,725446, load: 1,646205
```

**Comment by mickem** [comment]
Just a quick update, I havent abandoned this and after looking into it the new plan is to go "back to PDH" though that tends to be painful as PDH is localized. 

That said this week is a "winter holiday" in Sweden so I am away skiing (unfortunately not much snow) and wont have much time to tinker with that, so will take a while...

**Comment by mickem** [comment]
A slight updated. In preparation of adding pdh backed CPU load I have cleanup the existing code a bit. This means you should no longer get large negativ values instead you should get 0.

So if someone could give this a spin and let me know if thats the case that would be great:  [Version 0.6.13 can be download here](https://github.com/mickem/nscp/releases/tag/0.6.13)

**Comment by dneuhaeuser** [comment]
have v0.6.13 running for 24h now: 
no more large/negative values, but it is reporting 0 most of the time (while taskmanager is showing 4-5%)

**Comment by mickem** [comment]
[0.6.14](https://github.com/mickem/nscp/releases/tag/0.6.14) has experimental support for switch to PDH.

Add the following config:
```
[/settings/system/windows]
use pdh for cpu=true
```

If you check the debug log you can verify that this is enabled during startup:
```
D  w32system Loading counter: cpu_kernel = \2610(_Total)\144
D  w32system Loading counter: cpu_user = \2610(_Total)\142
```

These are the two counters used to measure CPU load:

* \Processor Information(_Total)\% Privileged Time for kernel/system load
* \Processor Information(_Total)\% User Time for user load

Please do let me know both if this resolves the issue as well as if these values seem accurate.

// Michael Medin

**Comment by dneuhaeuser** [comment]
I'm getting two errors in log, probably because this system is installed with german language:

```
2025-03-23 17:42:24: error:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:147: Exception in Failed to add counter cpu_kernel: : \Sitzungen in Verwendung(_Total)\Privilegierte Zeit (%) counter not found: c0000bb8: Das angegebene Objekt wurde nicht auf dem Computer gefunden.
2025-03-23 17:42:24: error:D:\a\nscp\nscp\modules\CheckSystem\pdh_thread.cpp:147: Exception in Failed to add counter cpu_user: : \Sitzungen in Verwendung(_Total)\Benutzerzeit (%) counter not found: c0000bb8: Das angegebene Objekt wurde nicht auf dem Computer gefunden.

```

**Comment by mickem** [comment]
Off the top of my mind maybe _Total is also translated and there seems to be no index for that so I shall look into that a bit.
It is also possible that there are multiple counters in Swedish and I accidentally got the wrong index 

You can manually override the counters used by adding the following config:

```ini
[/settings/system/windows/counters/cpu_user]
counter=\Sitzungen in Verwendung(_Total)\Benutzerzeit (%)
collection strategy=rrd

[/settings/system/windows/counters/cpu_kernel]
counter=\Sitzungen in Verwendung(_Total)\Privilegierte Zeit (%)
collection strategy=rrd
```

English names are either:
* Processor Information\% Processor Utility
* Processor Information\% Privileged Utility
or:
* Processor Information\% Processor Time
* Processor Information\% Privileged Time



To find the counter names in german you can run:
```
nscp sys --list --all
```

**Comment by dneuhaeuser** [comment]
```
counter=\Sitzungen in Verwendung(_Total)\Benutzerzeit (%)
counter=\Sitzungen in Verwendung(_Total)\Privilegierte Zeit (%)

```
Ok, these actually weren't the right counter names.

I think the right german translations are: 
```
Processor Information\% Processor Utility  =  Prozessorinformationen(_Total)\Prozessorauslastung
Processor Information\% Privileged Utility  =  Prozessorinformationen(_Total)\Privilegierte Auslastung
```

but cpu_user and cpu_kernel are added up in nscp, right?

"Processor Utility" is the total usage (corresponds exactly to taskmanager)

I cannot find a counter for something like "User Utility"...
only "User Time (%)", but that values do not seem to fit in...

**Comment by mickem** [comment]
New attempt with uses English counter names instead of indexes. It also adds support for cores (i.e. checking load per core) and I hope the total load.

Check it out here: https://github.com/mickem/nscp/releases/tag/0.6.15

Also if you overridode the counters you should probably remove that and ensure the default ones works.

So remove this:

```ini
[/settings/system/windows/counters/cpu_user]
...
[/settings/system/windows/counters/cpu_kernel]
...
```

But keep this:

```ni
[/settings/system/windows]
use pdh for cpu=true
```

**Comment by dneuhaeuser** [comment]
OK, I removed the overrides, kept "use pdh..." and have v0.6.15 running for a few hours now:

no more errors and the values now perfectly match those in taskmanager.

just out of curiosity: which counters are used now?

**Comment by mickem** [comment]
That is great news! 

I'll push a new release (i.e. no preview build) with a feature version bump over the weekend with this (and all other changes).
So please do let me know if anything changes...

The counters used are:
* \Processor Information($INSTANCE$)\% Processor Utility
* \Processor Information($INSTANCE$)\% Privileged Utility
And the math is:
* kernel = % Privileged Utility
* user = % Processor Utility - % Privileged Utility
* idle = 100 - % Processor Utility

// Michael Medin

**Comment by mickem** [comment]
Closing this as 0.7.0 is released now... If anyone still have issues please reopen.
